go-controller/pkg/node/secondary_node_network_controller_test.go

```diff
@@ -58,14 +62,89 @@ var _ = Describe("SecondaryNodeNetworkController", func() {
 		controller := SecondaryNodeNetworkController{}
 		var err error
 		controller.watchFactory, err = factory.NewNodeWatchFactory(fakeClient, "worker1")
+		Expect(err).NotTo(HaveOccurred())
 		Expect(controller.watchFactory.Start()).To(Succeed())
 
-		Expect(err).NotTo(HaveOccurred())
 		controller.NetInfo, err = util.ParseNADInfo(nad)
 		Expect(err).NotTo(HaveOccurred())
 
 		networkID, err := controller.getNetworkID()
 		Expect(err).To(HaveOccurred())
 		Expect(networkID).To(Equal(util.InvalidNetworkID))
 	})
+	It("ensure UDNGateway is not invoked when feature gate is OFF", func() {
+		config.OVNKubernetesFeature.EnableNetworkSegmentation = false
+		config.OVNKubernetesFeature.EnableMultiNetwork = true
+		factoryMock := factoryMocks.NodeWatchFactory{}
+		nodeList := []*corev1.Node{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "worker1",
+					Annotations: map[string]string{
+						"k8s.ovn.org/network-ids": `{"bluenet": "3"}`,
+					},
+				},
+			},
+		}
+		cnnci := CommonNodeNetworkControllerInfo{name: "worker1", watchFactory: &factoryMock}
+		factoryMock.On("GetNode", "worker1").Return(nodeList[0], nil)
+		factoryMock.On("GetNodes").Return(nodeList, nil)
+		NetInfo, err := util.ParseNADInfo(nad)
+		Expect(err).NotTo(HaveOccurred())
+		controller := NewSecondaryNodeNetworkController(&cnnci, NetInfo)
+		err = controller.Start(context.Background())
+		Expect(err).NotTo(HaveOccurred())
+		Expect(controller.gateway).To(BeNil())
+	})
+	It("ensure UDNGateway is invoked for Primary UDNs when feature gate is ON", func() {
+		config.OVNKubernetesFeature.EnableNetworkSegmentation = true
+		config.OVNKubernetesFeature.EnableMultiNetwork = true
+		factoryMock := factoryMocks.NodeWatchFactory{}
+		nodeList := []*corev1.Node{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "worker1",
+					Annotations: map[string]string{
+						"k8s.ovn.org/network-ids": `{"bluenet": "3"}`,
+					},
+				},
+			},
+		}
+		cnnci := CommonNodeNetworkControllerInfo{name: "worker1", watchFactory: &factoryMock}
+		factoryMock.On("GetNode", "worker1").Return(nodeList[0], nil)
+		factoryMock.On("GetNodes").Return(nodeList, nil)
+		NetInfo, err := util.ParseNADInfo(nad)
+		Expect(err).NotTo(HaveOccurred())
+		controller := NewSecondaryNodeNetworkController(&cnnci, NetInfo)
+		err = controller.Start(context.Background())
+		Expect(err).To(HaveOccurred()) // we don't have the gateway pieces setup so its expected to fail here
+		Expect(err.Error()).To(ContainSubstring("no annotation found"))
+		Expect(controller.gateway).To(Not(BeNil()))
+	})
+	It("ensure UDNGateway is invoked for Primary UDNs when feature gate is ON but network is not Primary", func() {
```

Comment:
the title of this is confusing. It makes it sound like the mp0 should get created, but actually ou are checking it does not. Maybe like: "Ensures UDNGateway is not created for UDN when feature gate is ON but network is not Primary"

Reply:
yeah lol :) that was a copy-paste snafu, nice catch tim!